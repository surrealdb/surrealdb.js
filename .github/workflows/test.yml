name: test

on:
  push:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Setup node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18
      - name: Setup deno
        uses: denoland/setup-deno@v1
        with:
          deno-version: v1.x
      - name: Setup bun
        uses: antongolub/action-setup-bun@v1
        with:
          bun-version: 0.1.13

      - uses: actions/checkout@v3

      - name: Build NPM package
        run: deno task build

      - name: Start SurrealDB
        run: sh ./test/e2e/scripts/docker-restart.sh

      - name: Run Deno
        run: deno test -A --trace-ops ./test/e2e/deno.js
      - name: Run Node
        run: node ./test/e2e/node.js
      - name: Run Bun
        run: bun run ./test/e2e/bun.js

      - name: Stop SurrealDB
        run:  sh ./test/e2e/scripts/docker-stop.sh
