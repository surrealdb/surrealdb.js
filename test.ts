import { RecordId } from "./src/library/data/recordid.ts";
import { Duration } from "./src/library/data/duration.ts";
import { UUID } from "./src/library/data/uuid.ts";
import { Decimal } from "./src/library/data/decimal.ts";
import { Surreal } from "./src/index.ts";
import { encode } from "./src/library/data/cbor.ts";
import { GeometryCollection, GeometryLine, GeometryPoint } from "./src/library/data/geometry.ts";
// import { recordId } from "./src/library/orm/types.ts";

const surreal = new Surreal();
await surreal.connect('ws://127.0.0.1:8000/rpc');
await surreal.use({
	namespace: 'test',
	database: 'test',
});

// console.log(
// 	await surreal.query<[unknown[]]>(
// 		"[$d, $r, $c, $u, $dur, $none, NONE, $a, <string> $a, $nan, $geo]",
// 		{
// 			d: new Date(),
// 			r: new RecordId('person', 'tobie'),
// 			c: new RecordId('recording', [
// 				new Date(),
// 				'London',
// 				new Uuid(),
// 				{
// 					temparature: new Decimal("10.00003")
// 				}
// 			]),
// 			u: new Uuid(),
// 			dur: new Duration('1d20m'),
// 			none: undefined,
// 			a: 9223372036854775807n,
// 			nan: NaN,
// 			geo: new GeometryCollection([
// 				new GeometryPoint([1, 2]),
// 				new GeometryLine([ new GeometryPoint([1, 2]), new GeometryPoint([3, 4]) ])
// 			])
// 		}
// 	)
// )

const uuid = new Duration("2w6d100ms");
console.log(
	456,
	await surreal.query<[UUID, string]>(
		"$uuid; <string> $uuid",
		{
			uuid
		}
	).then(([u, a]) => [u.toString(), a])
)

// function bufferToHex (buffer: ArrayBuffer) {
//     return [...new Uint8Array (buffer)]
//         .map (b => b.toString (16).padStart (2, "0"))
//         .join ("");
// }

// console.log(bufferToHex(encode(new RecordId('recording', [
// 	new Date(),
// 	'London',
// 	new Uuid(),
// 	{
// 		temparature: new Decimal("10.00003")
// 	}
// ]))))

// const rid = recordId('test');
// console.log(rid.parse(new RecordId('tesst', 123) ))

// console.log()
