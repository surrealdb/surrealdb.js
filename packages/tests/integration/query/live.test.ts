import { describe, expect, mock, test } from "bun:test";
import { RecordId, type Uuid } from "surrealdb";
import { insertMockRecords, personTable, setupServer } from "../__helpers__";

const { createSurreal, kill, spawn } = await setupServer();

describe("live() / liveOf()", async () => {
	const surreal = await createSurreal({
		protocol: "ws",
		reconnect: {
			enabled: true,
		},
	});

	await insertMockRecords(surreal);

	test("subscription properties", async () => {
		const subscription = await surreal.live(personTable);

		expect(subscription.isAlive).toBeTrue();
		expect(subscription.isManaged).toBeTrue();
		expect(subscription.resource).toEqual(personTable);

		await subscription.kill();

		expect(subscription.isAlive).toBeFalse();
	});

	test("create action", async () => {
		const subscription = await surreal.live(personTable);
		const { promise, resolve } = Promise.withResolvers();
		const mockHandler = mock(() => resolve());

		subscription.subscribe(mockHandler);

		await surreal.create(new RecordId("person", 3), {
			firstname: "John",
			lastname: "Doe",
		});

		await promise;

		expect(mockHandler).toBeCalledTimes(1);
		expect(mockHandler).toBeCalledWith(
			"CREATE",
			{
				id: new RecordId("person", 3),
				firstname: "John",
				lastname: "Doe",
			},
			new RecordId("person", 3),
		);

		await subscription.kill();
	});

	test("update action", async () => {
		const subscription = await surreal.live(personTable);
		const { promise, resolve } = Promise.withResolvers();
		const mockHandler = mock(() => resolve());

		subscription.subscribe(mockHandler);

		await surreal.update(new RecordId("person", 3), {
			firstname: "John",
			lastname: "Doe",
			age: 20,
		});

		await promise;

		expect(mockHandler).toBeCalledTimes(1);
		expect(mockHandler).toBeCalledWith(
			"UPDATE",
			{
				id: new RecordId("person", 3),
				firstname: "John",
				lastname: "Doe",
				age: 20,
			},
			new RecordId("person", 3),
		);

		await subscription.kill();
	});

	test("delete action", async () => {
		const subscription = await surreal.live(personTable);
		const { promise, resolve } = Promise.withResolvers();
		const mockHandler = mock(() => resolve());

		subscription.subscribe(mockHandler);

		await surreal.delete(new RecordId("person", 3));

		await promise;

		expect(mockHandler).toBeCalledTimes(1);
		expect(mockHandler).toBeCalledWith(
			"DELETE",
			{
				id: new RecordId("person", 3),
				firstname: "John",
				lastname: "Doe",
				age: 20,
			},
			new RecordId("person", 3),
		);

		await subscription.kill();
	});

	test("reconnect and resume", async () => {
		const subscription = await surreal.live(personTable);
		const { promise, resolve } = Promise.withResolvers();
		const mockHandler = mock(() => resolve());

		subscription.subscribe(mockHandler);

		const initialId = subscription.id;

		// Restart server and wait for reconnection
		await kill();
		await spawn();
		await surreal.ready;

		// Make sure we obtained a new live id
		expect(initialId).not.toEqual(subscription.id);

		await surreal.create(new RecordId("person", 3), {
			firstname: "John",
			lastname: "Doe",
		});

		await promise;

		expect(mockHandler).toBeCalledTimes(1);
		expect(mockHandler).toBeCalledWith(
			"CREATE",
			{
				id: new RecordId("person", 3),
				firstname: "John",
				lastname: "Doe",
			},
			new RecordId("person", 3),
		);

		await subscription.kill();

		expect(subscription.isAlive).toBeFalse();
	});

	test("unmanaged subscription properties", async () => {
		const [liveId] = await surreal.query<[Uuid]>("LIVE SELECT * FROM person");
		const subscription = await surreal.liveOf(liveId);

		expect(subscription.isAlive).toBeTrue();
		expect(subscription.isManaged).toBeFalse();
		expect(subscription.resource).toBeUndefined();

		await subscription.kill();

		expect(subscription.isAlive).toBeFalse();
	});

	test("unmanaged create action", async () => {
		const [liveId] = await surreal.query<[Uuid]>("LIVE SELECT * FROM person");
		const subscription = await surreal.liveOf(liveId);
		const { promise, resolve } = Promise.withResolvers();
		const mockHandler = mock(() => resolve());

		subscription.subscribe(mockHandler);

		await surreal.create(new RecordId("person", 4), {
			firstname: "John",
			lastname: "Doe",
		});

		await promise;

		expect(mockHandler).toBeCalledTimes(1);
		expect(mockHandler).toBeCalledWith(
			"CREATE",
			{
				id: new RecordId("person", 4),
				firstname: "John",
				lastname: "Doe",
			},
			new RecordId("person", 4),
		);

		await subscription.kill();
	});

	test("iterator", async () => {
		const subscription = await surreal.live(personTable);
		const iterator = subscription.iterate();

		await surreal.create(new RecordId("person", 5), {
			firstname: "John",
			lastname: "Doe",
		});

		await surreal.merge(new RecordId("person", 5), {
			firstname: "Mary",
		});

		await surreal.delete(new RecordId("person", 5));

		expect(await iterator.next()).toEqual({
			done: false,
			value: [
				"CREATE",
				{
					id: new RecordId("person", 5),
					firstname: "John",
					lastname: "Doe",
				},
				new RecordId("person", 5),
			],
		});
		expect(await iterator.next()).toEqual({
			done: false,
			value: [
				"UPDATE",
				{
					id: new RecordId("person", 5),
					firstname: "Mary",
					lastname: "Doe",
				},
				new RecordId("person", 5),
			],
		});
		expect(await iterator.next()).toEqual({
			done: false,
			value: [
				"DELETE",
				{
					id: new RecordId("person", 5),
					firstname: "Mary",
					lastname: "Doe",
				},
				new RecordId("person", 5),
			],
		});

		await subscription.kill();

		expect(await iterator.next()).toEqual({
			done: true,
			value: ["CLOSED", "KILLED"],
		});
	});

	test("iterator survives reconnect", async () => {
		const subscription = await surreal.live(personTable);
		const iterator = subscription.iterate();
		const initialId = subscription.id;

		// Restart server and wait for reconnection
		await kill();
		await spawn();
		await surreal.ready;

		// Make sure we obtained a new live id
		expect(initialId).not.toEqual(subscription.id);

		await surreal.create(new RecordId("person", 6), {
			firstname: "John",
			lastname: "Doe",
		});

		expect(await iterator.next()).toEqual({
			done: false,
			value: [
				"CREATE",
				{
					id: new RecordId("person", 6),
					firstname: "John",
					lastname: "Doe",
				},
				new RecordId("person", 6),
			],
		});

		subscription.kill();

		expect(await iterator.next()).toEqual({
			done: true,
			value: ["CLOSED", "KILLED"],
		});

		expect(subscription.isAlive).toBeFalse();
	});
});
