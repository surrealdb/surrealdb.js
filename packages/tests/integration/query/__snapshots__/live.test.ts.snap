// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`live() / liveOf() compile 1`] = `"LIVE SELECT type::fields($bind__1) FROM $bind__2 WHERE age = $bind__3 FETCH type::fields($bind__4)"`;

exports[`live() / liveOf() compile 2`] = `
{
  "bind__1": [
    "age",
    "test",
    "lastname",
  ],
  "bind__2": Table {},
  "bind__3": 30,
  "bind__4": [
    "foo",
  ],
}
`;
